---
title: 'Hepatoprotective effects of systemic ER activation'
subtitle: 'ChIPseq/Epigenome genome - Differential binding analysis'
author: 'Christian Sommerauer & Carlos Gallardo'
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    highlight: tango
---

```{r setup, include=FALSE}
# set path
path <- strsplit(dirname(rstudioapi::getActiveDocumentContext()$path), '/')[[1]]
path <- path[-c((length(path)-1):length(path))]
path <- paste(path, collapse = '/')
knitr::opts_knit$set(root.dir = path)
knitr::opts_chunk$set(echo = T, message = F, warning = F, dev = 'png')
```

```{r}
library(tidyverse)
library(DiffBind)
```

```{r}
sessionInfo()
```

```{r}
getwd()
samples <- c("CD2", "CD9", "HFD3", "HFD4","DPN2", "DPN3", "E2_8", "E2_9")
Factor_K27 <- c(rep("H3K27ac", 8))
PeakCaller <- c(rep("macs", 8))
Replicate <- c(rep(1:2, 4))
Treatment <- c(rep("CD", 2), rep("HFD", 2), rep("DPN", 2), rep("E2", 2)) 
files_K27 <- c("data/Downloaded_from_Arrayexpress/Peakfiles/CK0744_H3K27ac_CD2_peaks.broadPeak",
           "data/Downloaded_from_Arrayexpress/Peakfiles/CK0745_H3K27ac_CD9_peaks.broadPeak",
           "data/Downloaded_from_Arrayexpress/Peakfiles/CK0746_H3K27ac_HFD3_peaks.broadPeak",
           "data/Downloaded_from_Arrayexpress/Peakfiles/CK0747_H3K27ac_HFD4_peaks.broadPeak",
           "data/Downloaded_from_Arrayexpress/Peakfiles/CK0748_H3K27ac_DPN2_peaks.broadPeak",
           "data/Downloaded_from_Arrayexpress/Peakfiles/CK0749_H3K27ac_DPN3_peaks.broadPeak",
           "data/Downloaded_from_Arrayexpress/Peakfiles/CK0750_H3K27ac_E2_8_peaks.broadPeak",
           "data/Downloaded_from_Arrayexpress/Peakfiles/CK0751_H3K27ac_E2_9_peaks.broadPeak")

bam.files_K27 <- c("data/Downloaded_from_Arrayexpress/Bamfiles/CK0744_H3K27ac_CD2_psort_BL_fix_MkDup.bam",
               "data/Downloaded_from_Arrayexpress/Bamfiles/CK0745_H3K27ac_CD9_psort_BL_fix_MkDup.bam",
               "data/Downloaded_from_Arrayexpress/Bamfiles/CK0746_H3K27ac_HFD3_psort_BL_fix_MkDup.bam",
               "data/Downloaded_from_Arrayexpress/Bamfiles/CK0747_H3K27ac_HFD4_psort_BL_fix_MkDup.bam",
               "data/Downloaded_from_Arrayexpress/Bamfiles/CK0748_H3K27ac_DPN2_psort_BL_fix_MkDup.bam",
               "data/Downloaded_from_Arrayexpress/Bamfiles/CK0749_H3K27ac_DPN3_psort_BL_fix_MkDup.bam",
               "data/Downloaded_from_Arrayexpress/Bamfiles/CK0750_H3K27ac_E2_8_psort_BL_fix_MkDup.bam",
               "data/Downloaded_from_Arrayexpress/Bamfiles/CK0751_H3K27ac_E2_9_psort_BL_fix_MkDup.bam")

metaData_K27 <- data.frame(SampleID = samples, Factor = Factor_K27, Replicate = Replicate, Peaks = files_K27, bamReads = bam.files_K27, PeakCaller, Treatment=Treatment)
samplesheet_K27 <- dba(sampleSheet=metaData_K27)
```
```{r}
dba.plotHeatmap(samplesheet_K27)
```

```{r}
samplesheet.counted_K27 <- dba.count(samplesheet_K27, minOverlap = 2)
dba.overlap(samplesheet_K27,mode=DBA_OLAP_RATE)
```

```{r}
#That is a PCA based on read counts
dba.plotPCA(samplesheet.counted_K27, label=DBA_ID)
```

```{r}
contrasts_K27 <- dba.contrast(samplesheet.counted_K27, categories=c(DBA_TREATMENT), minMembers=2)
samplesheet.analysed_K27 <- dba.analyze(contrasts_K27)
```

```{r}
#That is a PCA based on diffbound sites of given contrast
##Get contrast numbers
dba.show(samplesheet.analysed_K27, bContrast=T)
dba.plotPCA(samplesheet.analysed_K27, contrast = 1)
```

```{r}
dba.plotVolcano(samplesheet.analysed_K27, contrast = 1)
```

```{r}
dba.plotBox(samplesheet.analysed_K27, contrast =5)
```
# Choose appropriate contrasts, export the full tables, then filter for signficant changes with a minimum foldchange of 50% up or down. 
# Then split the peaks into Loss/Gain in HFD and export bed files with these sites.
```{r}
res_deseq_K27_CD_vs_HFD <- as.data.frame(dba.report(samplesheet.analysed_K27, method=DBA_DESEQ2, contrast = 1, th=1))
res_deseq_K27_HFD_vs_DPN <- as.data.frame(dba.report(samplesheet.analysed_K27, method=DBA_DESEQ2, contrast = 4, th=1))
res_deseq_K27_HFD_vs_E2 <- as.data.frame(dba.report(samplesheet.analysed_K27, method=DBA_DESEQ2, contrast = 5, th=1))
res_deseq_K27_E2_vs_DPN <- as.data.frame(dba.report(samplesheet.analysed_K27, method=DBA_DESEQ2, contrast = 6, th=1)) # This comparison will not be needed further

Diffbind_tables <- list()
Diffbind_tables$unfiltered <- list(K27_CDvsHFD = res_deseq_K27_CD_vs_HFD,
                        K27_HFDvsDPN = res_deseq_K27_HFD_vs_DPN,
                        K27_HFDvsE2 = res_deseq_K27_HFD_vs_E2)

Diffbind_tables$all_DB_peaks <- list()
Diffbind_tables$up_DB_peaks <- list()
Diffbind_tables$down_DB_peaks <- list()

Diffbind_bed_files <- list(
  all_DB_peaks=list(),
  up_DB_peaks=list(),
  down_DB_peaks=list())

# This loop saves the diffbound peaks in the diffbind_tables list
# This loop also saves the bedfiles in a separate list

for (i in 1:length(Diffbind_tables$unfiltered)) { 
  
  ### PART 1 - Put the up- and downregulated peak objects (in the given contrast orientations) into dataframes
  #Both up-and downregulated together
  Diffbind_tables$all_DB_peaks[[i]] <- Diffbind_tables$unfiltered[[i]] %>% dplyr::filter(FDR< 0.05 & abs(Fold) > 0.585)
  #Only the upregulated peaks 
  Diffbind_tables$up_DB_peaks[[i]] <- Diffbind_tables$unfiltered[[i]] %>% dplyr::filter(FDR< 0.05 & Fold > 0.585)
  #Only the downregulated peaks 
  Diffbind_tables$down_DB_peaks[[i]] <- Diffbind_tables$unfiltered[[i]] %>% dplyr::filter(FDR< 0.05 & Fold < -0.585)
  
  ### PART 2 - Convert the objects to bedgraph format.
  Diffbind_bed_files$all_DB_peaks[[i]] <- Diffbind_tables$unfiltered[[i]] %>%
    dplyr::filter(FDR< 0.05 & abs(Fold) > 0.585) %>%
    dplyr::select(1:3) %>% mutate(name = "NA") %>% mutate(strand = "NA") 
  
  Diffbind_bed_files$up_DB_peaks[[i]] <- Diffbind_tables$unfiltered[[i]] %>%
    dplyr::filter(FDR< 0.05 & Fold > 0.585) %>% 
    dplyr::select(1:3) %>% mutate(name = "NA") %>% mutate(strand = "NA") 
  
  Diffbind_bed_files$down_DB_peaks[[i]] <- Diffbind_tables$unfiltered[[i]] %>%
    dplyr::filter(FDR< 0.05 & Fold < -0.585) %>% 
    dplyr::select(1:3) %>% mutate(name = "NA") %>% mutate(strand = "NA") 
}

names(Diffbind_tables$all_DB_peaks) <- names(Diffbind_tables$unfiltered)
names(Diffbind_tables$up_DB_peaks) <- names(Diffbind_tables$unfiltered)
names(Diffbind_tables$down_DB_peaks) <- names(Diffbind_tables$unfiltered)
names(Diffbind_bed_files$all_DB_peaks) <- names(Diffbind_tables$unfiltered)
names(Diffbind_bed_files$up_DB_peaks) <- names(Diffbind_tables$unfiltered)
names(Diffbind_bed_files$down_DB_peaks) <- names(Diffbind_tables$unfiltered)

saveRDS(Diffbind_tables, "results/Epigenome_analysis/Diffbind_results_FDR_fold.rds")

# Export the bedfiles. These will be used later in the analysis.

for (i in 1:length(Diffbind_bed_files)) {
  
write.table(x = Diffbind_bed_files$all_DB_peaks[[i]], file=paste0("results/Epigenome_analysis/Diffbind_all_DB_regions_broad_",
                                 names(Diffbind_bed_files$all_DB_peaks[i]),".bed"), sep="\t", quote=F, row.names=F, col.names=F)  
  
write.table(x = Diffbind_bed_files$up_DB_peaks[[i]], file=paste0("results/Epigenome_analysis/Diffbind_up_DB_regions_broad_",
                                 names(Diffbind_bed_files$up_DB_peaks[i]),".bed"), sep="\t", quote=F, row.names=F, col.names=F)

write.table(x = Diffbind_bed_files$down_DB_peaks[[i]], file=paste0("results/Epigenome_analysis/Diffbind_down_DB_regions_broad_",
                                 names(Diffbind_bed_files$down_DB_peaks[i]),".bed"), sep="\t", quote=F, row.names=F, col.names=F) 
}
```

#load the packages for the annotations
```{r}
library("ChIPpeakAnno")
sessionInfo()
library("GenomicRanges")
options(connectionObserver = NULL) #That is a work-around, as the org.Mm. package cannot be loaded 
library("org.Mm.eg.db")
library("biomaRt")
library( reactome.db )
```

#Annotate all of the coordinates of DiffBound peaks in the different objects. Nearest gene.
```{r}
# For annotation, we use the September 2019 version, which was also used for RNAseq.
listEnsemblArchives()
mart <- useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl", host = "https://sep2019.archive.ensembl.org")
# use the getAnnotation function to obtain the TSS for ensembl GRCh38.p13.
annoData <- getAnnotation(mart, featureType = "TSS")

# Annotate the peak files.
Diffbind_tables$annotated <- Diffbind_tables

for (i in 1:length(Diffbind_tables$annotated)) {
  for (k in 1:length(Diffbind_tables$annotated$unfiltered)) {

    colnames(Diffbind_tables$annotated[[i]][[k]]) <- c("chrom", "start", "end")
    nrow(Diffbind_tables$annotated[[i]][[k]])

Diffbind_tables$annotated[[i]][[k]] <- makeGRangesFromDataFrame(Diffbind_tables$annotated[[i]][[k]], 
                                                                start.field = "start", end.field = "end", ignore.strand = T) 

#Give ranges numeric names in order
names(Diffbind_tables$annotated[[i]][[k]]) <- c(1:length(Diffbind_tables$annotated[[i]][[k]]))

#Annotate granges with the nearest TSS
Diffbind_tables$annotated[[i]][[k]] <- annotatePeakInBatch(Diffbind_tables$annotated[[i]][[k]], 
                               AnnotationData=annoData, 
                               featureType = "TSS",
                               output="nearestLocation",
                               PeakLocForDistance = "start")

Diffbind_tables$annotated[[i]][[k]] <- as.data.frame(Diffbind_tables$annotated[[i]][[k]]) %>% remove_rownames()

  } 
}

saveRDS(Diffbind_tables$annotated, file="results/Epigenome_analysis/Diffbind_all_annotated_DB_peaks.rds")
```

```{r}
sessionInfo()
```


