---
title: 'Hepatoprotective effects of systemic ER activation'
subtitle: 'ChIPseq/Epigenome genome - Enhancer-gene pair analysis'
author: 'Christian Sommerauer & Carlos Gallardo'
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    highlight: tango
---

```{r setup, include=FALSE}
# set path
path <- strsplit(dirname(rstudioapi::getActiveDocumentContext()$path), '/')[[1]]
path <- path[-c((length(path)-1):length(path))]
path <- paste(path, collapse = '/')
knitr::opts_knit$set(root.dir = path)
knitr::opts_chunk$set(echo = T, message = F, warning = F, dev = 'png')
```

```{r}
library(tidyverse)
```

# Use BEDOPS suite to determine the closest TSS to the enhancer sites. Run this in terminal, may adjust paths. Requires BEDOPS and bedtools.

```{r}
#Import the closest genes as determined by 
closest <- read.delim("results/Epigenome_analysis/origin_closest1_left_closest_1_2_right.bed", header=F)
enhancer_ID <- c(1:1816)
# Row 738 has an NA in several columns (including V8), remove this one as it causes issues downstream. 
closest2 <- closest %>%
separate(col = V3, into = c("V3_left", "V3_right"), sep = "\\|") %>%
separate(col = V5, into = c("V5_left", "V5_right"), sep = "\\|") %>%
separate(col = V7, into = c("V7_left", "V7_right"), sep = "\\|") %>%
mutate(enhancer_ID=enhancer_ID) %>%
filter(!is.na(V8)) 

closest3 <- closest2 %>% mutate(enha.ident = paste0(V1, ".",V2,".",V3_left), .before = V1)

colnames(closest3) <- c("enha.ident", "ori_chrom", "ori_start", "ori_end", "left_1_chrom", "left_1_start", "left_1_end",
                       "right_1_chrom", "right_1_start", "right_1_end",
                       "right_2_chrom", "right_2_start", "right_2_end", "enhancer_ID")


  closest_ori <- closest3 %>% dplyr::select(1, 14, 2:4) %>% mutate(query_ID = "closest_ori")
  colnames(closest_ori) <- c("loc_ID", "enha.ID",  "chrom", "start", "end", "query_ID") 
  closest_left1 <- closest3 %>% dplyr::select(1, 14, 5:7) %>% mutate(query_ID = "closest_left1")
  colnames(closest_left1) <- c("loc_ID", "enha.ID",  "chrom", "start", "end", "query_ID") 
 
  closest_right1 <- closest3 %>% dplyr::select(1, 14, 8:10) %>% mutate(query_ID = "closest_right1")
  colnames(closest_right1) <- c("loc_ID", "enha.ID",  "chrom", "start", "end", "query_ID") 
  closest_right2 <- closest3 %>% dplyr::select(1, 14, 11:13) %>% mutate(query_ID = "closest_right2")
  colnames(closest_right2) <-  c("loc_ID", "enha.ID",  "chrom", "start", "end", "query_ID") 

closest_long <- rbind(closest_ori, closest_left1, 
                      closest_right1, closest_right2)

#subset the enhancer_df to have the exact locations in three columns for later
location <- closest_ori %>% dplyr::select(1,3,4,5)
```

# Annotate the closest genes
```{r}
library("ChIPpeakAnno")
library("GenomicRanges")
options(connectionObserver = NULL) #That is a work-around, as the org.Mm. package cannot be loaded 
library("org.Mm.eg.db")
library("biomaRt")

gr_closest_long <- makeGRangesFromDataFrame(closest_long,  start.field = "start", end.field = "end", ignore.strand = T,keep.extra.columns = T) 
names(gr_closest_long) <- c(1:length(gr_closest_long))
```
## Annotate the TSS
```{r}
listEnsemblArchives()
mart <- useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl", host = "https://sep2019.archive.ensembl.org")
annoDataMart <- getAnnotation(mart, featureType = "TSS")
```

## Annotate the TSS
```{r}
gr_closest_long_anno <- annotatePeakInBatch(gr_closest_long, 
                               AnnotationData=annoDataMart, 
                               featureType = "TSS",
                               output="nearestLocation",
                               PeakLocForDistance = "start")

gr_closest_long_anno <- as.data.frame(gr_closest_long_anno)
```

# Import the gene expression data
```{r}
getwd()

source("code/00_helper_functions.R")
symbol_geneID <- read.delim("data/ensembl_mmus_sep2019_annotation.tsv")[,1:2]

raw_counts <- read.table(
  file = 'data/bulkRNAseq_mmus_rawcounts.tsv',
  stringsAsFactors = FALSE,
  sep = '\t',
  header = TRUE) %>%
  dplyr::select(-PPT_HFD_male_4) %>% 
  tibble::column_to_rownames('geneID') %>% 
  as.matrix() 

gene_len <- read.table(
  file = 'data/bulkRNAseq_mmus_gene_lengths.tsv',
  stringsAsFactors = FALSE,
  sep = '\t',
  header = TRUE)

TPM <- normalizeData(x=raw_counts, len = gene_len$length, method = "TPM") %>%
  tibble::rownames_to_column("ensembl_gene_id")

TPM <- TPM %>%
  dplyr::select(ensembl_gene_id, CD_male_1, CD_male_4, HFD_male_2,HFD_male_1,DPN_HFD_male_1, DPN_HFD_male_2, E2_HFD_male_3,E2_HFD_male_4)
TPM <- inner_join(symbol_geneID, TPM, by="ensembl_gene_id")

# We name the mice according to their original mouse number instead of replicate number. 
# CD2 and CD9 correspond to CDm1 and CDm4, HFD3 and HFD4 correspond to HFDm2 and HFDm1, DPN2 and DPN3 correspond to DPN1 and DPN2 and E2_8/E2_9 correspond to E2_3 and E2_4.

colnames(TPM) <- c("ensembl_gene_id", "symbol", "CDm2", "CDm9", "HFDm3", "HFDm4","DPN2", "DPN3", "E2_8", "E2_9")

gr_closest_long_anno_closest_genes <- gr_closest_long_anno %>%
  filter(!query_ID=="closest_ori") %>%
  dplyr::rename("ensembl_gene_id"="feature")

gr_closest_long_anno_closest_genes <- as.data.frame(gr_closest_long_anno_closest_genes)

TPM_filt <- TPM %>%
  dplyr::filter(ensembl_gene_id%in%gr_closest_long_anno_closest_genes$ensembl_gene_id)

chrom_TPM <- inner_join(gr_closest_long_anno_closest_genes, TPM_filt, by= "ensembl_gene_id") # 2 entries get missing

chrom_TPM2 <- chrom_TPM %>%
  dplyr::select("loc_ID" ,"seqnames", "start", "end", "enha.ID", "query_ID", "symbol", "ensembl_gene_id", "CDm2", "CDm9", "HFDm3", "HFDm4","DPN2", "DPN3", "E2_8", "E2_9")
```
# IMPORT ENHANCER COUNTS and normalize table
```{r}
library(dplyr)
library(tidyr)

counts_enha <- read.delim("results/Epigenome_analysis/diffbind_enhancers_1816_H3K27ac.clean.readCount", header=T) %>% tibble::column_to_rownames("Geneid")
names(counts_enha) <- c("CDm2_K27ac","CDm9_K27ac","HFDm3_K27ac","HFDm4_K27ac", "DPN2_K27ac","DPN3_K27ac","E2_8_K27ac","E2_9_K27ac")
colsums_enha <- colSums(counts_enha[,])

counts_enha_norm <- sweep(counts_enha, 2, colsums_enha, FUN = "/") 
counts_enha_norm2 <- counts_enha_norm *10^6
colSums(counts_enha_norm2[,])

counts_enha_norm2.1 <- counts_enha_norm2 %>% rownames_to_column("loc_ID")
K27_GE_joined <- inner_join(chrom_TPM2, counts_enha_norm2.1, by="loc_ID")
View(K27_GE_joined)
```

## Subset the enhancer table and put into long format
```{r}
sub_GE.K27_GE_joined <- K27_GE_joined %>%
  dplyr::select("loc_ID", "query_ID","symbol", "CDm2","CDm9","HFDm3","HFDm4", "DPN2","DPN3","E2_8","E2_9")
sub_GE.K27_GE_long <- pivot_longer(sub_GE.K27_GE_joined, cols=4:11, values_to = "Gene_expression")

sub.K27_K27_GE_joined <- K27_GE_joined %>%
  dplyr::select("loc_ID", "query_ID", "ensembl_gene_id","CDm2_K27ac","CDm9_K27ac","HFDm3_K27ac","HFDm4_K27ac", "DPN2_K27ac","DPN3_K27ac","E2_8_K27ac","E2_9_K27ac")
sub.K27_K27_GE_long <- pivot_longer(sub.K27_K27_GE_joined, cols=4:11, values_to = "H3K27ac")

K27_GE_long <- cbind(sub_GE.K27_GE_long, sub.K27_K27_GE_long)

K27_GE_long_dd <- K27_GE_long[!duplicated(as.list(K27_GE_long))]

# Remove the zeros to not correlate zeros (gives error message - but these genes are removed later anyhow because they are not expressed)
K27_GE_long_dd <- K27_GE_long_dd %>%
  group_by(loc_ID, symbol) %>%
  mutate(filter_zeros = mean(Gene_expression)) %>%
  filter(filter_zeros > 0) %>%
  dplyr::select(!filter_zeros)
```

# Import the reverted gene sets and filter the tables
```{r}
K27_GE_long_group <- K27_GE_long_dd %>% group_by(loc_ID, query_ID) %>%
  mutate(correlation_pearson = cor(Gene_expression, H3K27ac, method="pearson")) %>%
  mutate(correlation_spearman = cor(Gene_expression, H3K27ac, method="spearman"))

write.table(K27_GE_long_group, "results/Epigenome_analysis/K27_GE_corr_before_filtering.txt", quote=F, row.names=F, sep="\t")

DEGsets <- readRDS("results/bulkRNAseq_mmus_DEG_sets.rds")
revALL <- DEGsets$gene_id$reverted
length(revALL)

K27_GE_long_rev_insect <-  K27_GE_long_group %>%
  filter(ensembl_gene_id %in% revALL)
length(unique(K27_GE_long_rev_insect$ensembl_gene_id))


K27_GE_long_group_plot_pearson <- K27_GE_long_rev_insect %>%
  filter(abs(correlation_pearson) > 0.75) %>%
  group_by(loc_ID, query_ID) %>%
  mutate(name.ident = paste0(symbol, "_", loc_ID))

K27_GE_long_group_plot_pearson_revCount <- K27_GE_long_group_plot_pearson %>%
  mutate(name.ident = paste0(loc_ID, ":", symbol))

K27_GE_long_group_plot_spearman <- K27_GE_long_rev_insect %>%
  filter(abs(correlation_spearman) > 0.75) %>% group_by(loc_ID, query_ID) %>%
  mutate(name.ident = paste0(symbol, "_", loc_ID))

K27_GE_long_group_plot_pearson_before_Rev <- K27_GE_long_group %>%
  filter(abs(correlation_pearson) > 0.75) %>% group_by(loc_ID, query_ID) %>%
  mutate(name.ident = paste0(symbol, "_", loc_ID))


K27_GE_long_group_plot_spearman_before_Rev <- K27_GE_long_group %>%
  filter(abs(correlation_spearman) > 0.75) %>% group_by(loc_ID, query_ID) %>%
  mutate(name.ident = paste0(symbol, "_", loc_ID))
```
#Add 50kb to intersect CTCF peaks with the H3K27ac peaks
```{r}
K27_GE_long_group_coordinates <- inner_join(K27_GE_long_group_plot_pearson, location, by="loc_ID")
K27_GE_long_group_coordinates$end <- as.integer(K27_GE_long_group_coordinates$end)

K27_GE_long_group_coordinates_left <- K27_GE_long_group_coordinates %>%
  dplyr::filter(query_ID=="closest_left1") %>%
  mutate(new_end = end+50000) %>%
  mutate(new_start=start)

K27_GE_long_group_coordinates_right <- K27_GE_long_group_coordinates %>%
  dplyr::filter(query_ID=="closest_right1" | query_ID=="closest_right2") %>%
  mutate(new_start = start-50000) %>%
  mutate(new_end=end)

K27_GE_long_group_coordinates_left_export <- K27_GE_long_group_coordinates_left %>%
  dplyr::select("chrom", "new_start", "new_end", "loc_ID", "query_ID", "symbol") %>% 
  unique()

K27_GE_long_group_coordinates_right_export <- K27_GE_long_group_coordinates_right %>%
  dplyr::select("chrom", "new_start", "new_end", "loc_ID", "query_ID", "symbol") %>%
  unique()

write.table(K27_GE_long_group_coordinates_left_export, "results/Epigenome_analysis/H3K27ac_left_non_intersect.bed", quote=F, row.names = F, sep="\t")
write.table(K27_GE_long_group_coordinates_right_export, "results/Epigenome_analysis/H3K27ac_right_non_intersect.bed", quote=F, row.names=F, sep="\t")

#From here, intersect the CTCF peaks with the exported H3K27ac regions using BEDtools (command line)
```

# prepare the CTCF files - separate by motif-orientation.
```{r}
#load the motif-discovery file of CTCF motifs in mm10 genome by FIMO
FIMO_CTCF <- read.delim("data/fimo_mm10_genome_CTCFscan.tsv", sep="\t")

FIMO_CTCF_plus_bed <- FIMO_CTCF %>%
  dplyr::filter(strand=="+") %>%
  dplyr::select("chrom"="sequence_name", "start", "end"="stop", "strand")

FIMO_CTCF_minus_bed <- FIMO_CTCF %>%
  dplyr::filter(strand=="-") %>%
  dplyr::select("chrom"="sequence_name", "start", "end"="stop", "strand")

write.table(FIMO_CTCF_plus_bed, "results/Epigenome_analysis/fimo_mm10_genome_CTCF_plus.bed", quote=F, row.names = F, sep="\t")
write.table(FIMO_CTCF_minus_bed, "results/Epigenome_analysis/fimo_mm10_genome_CTCF_minus.bed", quote=F, row.names = F, sep="\t")
```

###################################################################################################

HERE, RUN THE SHELL SCRIPT "Epigenome_06.03_CTCF_script_bedtools_enh_intersect.sh"

###################################################################################################

#after BEDTools intersection of H3K27ac enhancers (that have a good correlation with nearby genes) with nearby CTCF peaks, re-import
```{r}
library(tidyverse)

names <- c("chrom", "start", "end", "loc_ID", "query_ID", "symbol")
H3K27ac_left_CTCFx_outwards <- read.delim("results/Epigenome_analysis/H3K27ac_left_CTCF.intersect.noncanon.uniq.bed", header=F, col.names = names) %>% mutate(CTCF_pos = "outwards")

H3K27ac_left_CTCFx <- read.delim("results/Epigenome_analysis/H3K27ac_left_CTCF.intersect.canon.uniq.bed", header=F, col.names = names) %>% mutate(CTCF_pos = "canonical")

H3K27ac_right_CTCFx_outwards <- read.delim("results/Epigenome_analysis/H3K27ac_right_CTCF.intersect.noncanon.uniq.bed", header=F, col.names = names) %>% mutate(CTCF_pos = "outwards")

H3K27ac_right_CTCFx <- read.delim("results/Epigenome_analysis/H3K27ac_right_CTCF.intersect.canon.uniq.bed", header=F, col.names = names) %>% mutate(CTCF_pos = "canonical")
```


#combine these data.frames, because they comprise the enhancer-gene pairs that we can report
```{r}
H3K27ac_CTCF_intersect <- rbind(H3K27ac_left_CTCFx_outwards, H3K27ac_left_CTCFx, H3K27ac_right_CTCFx_outwards, H3K27ac_right_CTCFx)
table(H3K27ac_CTCF_intersect$CTCF_pos)
length(unique(H3K27ac_CTCF_intersect$loc_ID))
unique_symbols <- unique(H3K27ac_CTCF_intersect$symbol)
length(unique_symbols)

# 45 unique genes that underlie potential enhancer-mediated estrogen-dependent regulation
```

#Compare the fold-change values for these sites - in addition to the reads in peaks this gives information about how much these enhancers are changed
```{r}
CDvsHFD_H3K27ac_Diffbind <- readRDS("results/Epigenome_analysis/Diffbind_results_FDR_fold.rds")$all_DB_peaks$K27_CDvsHFD %>%
 mutate(loc_ID = paste0(seqnames, ".",start,".",end), .before = seqnames) %>%
  dplyr::select(loc_ID, Fold, FDR)

# These are the enhancers intersected with CTCF. But more informative with foldchanges from Diffbind. The correlations are not added here.
H3K27ac_CTCF_intersect_log2FC <- inner_join(H3K27ac_CTCF_intersect,CDvsHFD_H3K27ac_Diffbind, by="loc_ID") 

# Retrieves the unique person corr values from BEFORE the CTCF intersection 
corr_values <- K27_GE_long_group_plot_pearson %>%
  ungroup() %>%
  dplyr::select("loc_ID", "correlation_pearson", "symbol") %>%
  unique() %>% group_by(symbol)

# Creates a dataframe between the corr values and log2FCs. This one is the COMPLETE dataframe
H3K27ac_CTCF_intersect_log2FC_corr <- inner_join(corr_values, H3K27ac_CTCF_intersect_log2FC, by=c("loc_ID", "symbol"))
duplicated(H3K27ac_CTCF_intersect_log2FC_corr)
length(unique(H3K27ac_CTCF_intersect_log2FC_corr$symbol))

# This unique ID is necessary to join the data frames in the next section. Otherwise, enhancers that are associated with multiple genes but are only good regarding CTCF peaks with one of the genes due to directionality, will again introduce the genes without CTCF peak.
H3K27ac_CTCF_intersect_log2FC_corr <- H3K27ac_CTCF_intersect_log2FC_corr %>%
  mutate(name.ident = paste0(loc_ID, ":", symbol))
```

```{r}
# To plot, we need the single columns for gene expression and log2FC again.  Note: some enhancers have the same gene associated as closest AND second closest due to multiple 
K27_GE_long_group_plot_filt <- K27_GE_long_group_plot_pearson %>%
  group_by(symbol, loc_ID) %>%
  mutate(name.ident = paste0(loc_ID, ":", symbol)) %>%
  filter(name.ident%in%H3K27ac_CTCF_intersect_log2FC_corr$name.ident)

K27_GE_long_group_plot_filt <- K27_GE_long_group_plot_filt[!duplicated(K27_GE_long_group_plot_filt), ]
length(unique(K27_GE_long_group_plot_filt$symbol))

# The following should yield "character(0)"
setdiff(K27_GE_long_group_plot_filt$symbol, H3K27ac_CTCF_intersect_log2FC_corr$symbol)

length(unique(K27_GE_long_group_plot_filt$symbol)) # 45 unique genes
nrow(K27_GE_long_group_plot_filt)/8 # 68 enhancer - gene pairs
length(unique(K27_GE_long_group_plot_filt$loc_ID)) # 67 unique enhancers

# 68 total combinations of location IDs and genes, one enhancer goes for to genes.

write.table(K27_GE_long_group_plot_filt, "results/Epigenome_analysis/corr_45genes_67enh_toPlot.txt", row.names = F,  quote = F, sep="\t")
```

```{r}
library(ggpubr)
  ggscatter(K27_GE_long_group_plot_filt, x = "H3K27ac", y="Gene_expression", add = "reg.line",
            shape=21,size=5, fill="name",alpha=0.8,
            # yscale = "log2",
            # xscale = "log2",
            xlab="H3K27ac",
            ylab="TPM",
            palette=c("#7f9ad7","#7f9ad7", "#7dc7d1", "#7dc7d1", "#356fb5","#356fb5","#2c2f72","#2c2f72")) +
    stat_cor(aes(label = ..r.label..),method="pearson", p.digits=0, size=5) +
    theme_bw() + 
    theme(axis.text = element_text(size=14),
          strip.background = element_rect(colour="black",
                                        fill="white")) +
    facet_wrap(vars(symbol, loc_ID), scales = "free", ncol=8) 
  
Acot_filt <- K27_GE_long_group_plot_filt %>% dplyr::filter(grepl("Acot", symbol))
ggscatter(Acot_filt, x = "H3K27ac", y="Gene_expression", add = "reg.line",
            shape=21,size=5, fill="name",alpha=0.8,
            # yscale = "log2",
            # xscale = "log2",
            xlab="H3K27ac",
            ylab="TPM",
            palette=c("#7f9ad7","#7f9ad7", "#7dc7d1", "#7dc7d1", "#356fb5","#356fb5","#2c2f72","#2c2f72")) +
    stat_cor(aes(label = ..r.label..),method="pearson", p.digits=0, size=5) +
    theme_bw() + 
    theme(axis.text = element_text(size=14),
          strip.background = element_rect(colour="black",
                                        fill="white")) +
    facet_wrap(vars(symbol, loc_ID), scales="free", ncol=4) 


# plot Tead1
tead1 <- K27_GE_long_group %>% dplyr::filter(symbol=="Tead1") %>% filter(!loc_ID=="chr7.112688427.112688827")
# note: this loc_ID is filtered out because it does not have a CTCF peak closeby (at least not a significant one)
#tead1_filt <- K27_GE_long_group_plot_filt %>% dplyr::filter(symbol=="Tead1")  # Plot this to only show E-G-pairs > 0.75
ggscatter(tead1, x = "H3K27ac", y="Gene_expression", add = "reg.line",
            shape=21,size=5, fill="name",alpha=0.8,
            # yscale = "log2",
            # xscale = "log2",
            xlab="H3K27ac",
            ylab="TPM",
            palette=c("#7f9ad7","#7f9ad7", "#7dc7d1", "#7dc7d1", "#356fb5","#356fb5","#2c2f72","#2c2f72")) +
    stat_cor(aes(label = ..r.label..),method="pearson", p.digits=0, size=5) +
    theme_bw() + 
    theme(axis.text = element_text(size=14),
          strip.background = element_rect(colour="black",
                                        fill="white")) +
    facet_wrap(vars(symbol, loc_ID), scales="free", ncol=4) 

# One enhancer has a correlation of 0.67 and hence does not pass the filter. I think this enhancer is interesting to show, which is why I remove the two enhancers with really poor correlation in the next line, keeping three of five enhancers that we show.
Sulf2_also_non_filter_pass <- K27_GE_long_group %>% dplyr::filter(symbol=="Sulf2") %>% filter(!loc_ID=="chr2.166125461.166125861")  %>% filter(!loc_ID=="chr2.166676963.166677363")
Sulf2_all_enhancers <- K27_GE_long_group %>% dplyr::filter(symbol=="Sulf2")
#Sulf2 <- K27_GE_long_group_plot_filt %>% dplyr::filter(symbol=="Sulf2")  # Plot this to only show E-G-pairs > 0.75
ggscatter(Sulf2_also_non_filter_pass, x = "H3K27ac", y="Gene_expression", add = "reg.line",
            shape=21,size=5, fill="name",alpha=0.8,
            # yscale = "log2",
            # xscale = "log2",
            xlab="H3K27ac",
            ylab="TPM",
            palette=c("#7f9ad7","#7f9ad7", "#7dc7d1", "#7dc7d1", "#356fb5","#356fb5","#2c2f72","#2c2f72")) +
    stat_cor(aes(label = ..r.label..),method="pearson", p.digits=0, size=5) +
    theme_bw() + 
    theme(axis.text = element_text(size=14),
          strip.background = element_rect(colour="black",
                                        fill="white")) +
    facet_wrap(vars(symbol, loc_ID), scales="free", ncol=4) 
```

# Plot histogram to show in which processes (of the 24 GSEA) the 45 genes fall into.
```{r}
library(clusterProfiler)
K27_GE_long_group_plot_filt <- read.delim("results/Epigenome_analysis/corr_45genes_67enh_toPlot.txt")
length(unique(K27_GE_long_group_plot_filt$symbol))
symbols <- K27_GE_long_group_plot_filt$symbol %>% unique()

reactome_pathways <- readRDS("results/bulkRNAseq_mmus_GSEA_reactome_cluster_sets.rds")
reactome_pathways.2 <- as.data.frame(do.call(cbind, reactome_pathways))

reactome_pathways.long <- pivot_longer(reactome_pathways.2, cols=1:24)
intersect_pathways_symbols <- reactome_pathways.long %>% filter(value %in% symbols) %>% unique()
intersect_pathways_symbols_counts <- as.data.frame(table(intersect_pathways_symbols$name))

order.GSEA.pathways <- as.data.frame(table(intersect_pathways_symbols$name)) %>%
  arrange(-Freq) %>%
  dplyr::pull("Var1") %>%  
  as.vector()

str(order.GSEA.pathways)

ggplot(intersect_pathways_symbols) +
  geom_histogram(aes(x=factor(name, levels=rev(order.GSEA.pathways))), stat="count", fill="grey", color="black") +
  coord_flip() +
  theme_classic() +
  theme(axis.text.x = element_text(vjust = .5)) +
  xlab("") +
  ylab("Gene count") +
  scale_y_continuous(limits = c(), breaks=c(0,2,4,6,8,10)) 
```


```{r}
library(clusterProfiler)
options(connectionObserver = NULL) 
# Warning: call dbDisconnect() when finished working with a connection
library(org.Mm.eg.db)

unique_symbols <- K27_GE_long_group_plot_filt$symbol %>% unique()
length(unique_symbols)

GO_BP <- enrichGO(gene = unique_symbols,
                keyType       = 'SYMBOL',
                OrgDb         = org.Mm.eg.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                 pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                minGSSize     = 1,
                readable      = F,
                universe = TPM_filt$symbol) # Can also use all expressed genes here instead.
head(as.data.frame(GO_BP))
View(as.data.frame(GO_BP))

  library(dplyr)
  library(ggplot2)
  plot_me_ordered <- GO_BP[order(GO_BP$p.adjust), ]
  plot_me_ordered <- plot_me_ordered[1:12, ]
  plot_me_ordered <- plot_me_ordered[order(plot_me_ordered$Count), ]
  name_order <- plot_me_ordered %>%
    dplyr::pull("Description")
  
  ggplot(plot_me_ordered, aes(x=factor(Description, levels=name_order), fill=-log10(p.adjust), y=Count)) +
    geom_bar(stat="identity") +
    coord_flip() +
    scale_fill_gradient(low = "#808b96", high = "black", 
    limits = c(2, 6), breaks = c(2, 4, 6))+
    theme_classic() +
    theme(text=element_text(size = 18)) +
    ggtitle("") + 
    xlab("") 
```

```{r}
sessionInfo()
```

