---
title: 'Hepatoprotective effects of systemic ER activation'
subtitle: 'ChIPseq/Epigenome genome - Annotation of Peak files and feature examination'
author: 'Christian Sommerauer & Carlos Gallardo'
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    highlight: tango
---

```{r setup, include=FALSE}
# set path
path <- strsplit(dirname(rstudioapi::getActiveDocumentContext()$path), '/')[[1]]
path <- path[-c((length(path)-1):length(path))]
path <- paste(path, collapse = '/')
knitr::opts_knit$set(root.dir = path)
knitr::opts_chunk$set(echo = T, message = F, warning = F, dev = 'png')
```


# Load the enhancers and promoters which were demarcated using various histone modifications. Then annotate these features and export them as tables. For promoters, only one peak per gene, the closest one, is permitted, multiple TSS are disregaded.
```{r}
library("ChIPpeakAnno")
library("GenomicRanges")
options(connectionObserver = NULL) #That is a work-around, as the org.Mm. package cannot be loaded 
library("org.Mm.eg.db")
library("biomaRt")
library("tidyverse")

import_list <- list(
promoters_genomewide <- read.delim("results/Epigenome_analysis/Final_promoter_files/prom.3.genomewide.FINAL.H3K4me3_CD_HFD_merge.bed", header=F),
enhancers_genomewide <- read.delim("results/Epigenome_analysis/Final_enhancer_files/enh.5.FINAL.genomewide.H3K27ac_all_xxx_H3K4me1_minus_K4me3.bed", header=F),
enhancers_allDB <- read.delim("results/Epigenome_analysis/Final_enhancer_files/enh.5.FINAL.H3K27ac_broad_CDHFD_DB_xxx_H3K4me1_minus_me3.sort.bed", header=F),
promoters_allDB <- read.delim("results/Epigenome_analysis/Final_promoter_files/prom.8.FINAL.H3K27ac_broad_CDHFD_DB_xxx_H3K4me3_CD_HFD.bed", header=F),
enhancers_HFDup <- read.delim("results/Epigenome_analysis/Final_enhancer_files/enh.7.HFDup.FINAL.H3K27ac_broad_CDHFD_DB_xxx_H3K4me1_minus_me3.sort.bed", header=F),
promoters_HFDup <- read.delim("results/Epigenome_analysis/Final_promoter_files/prom.16.FINAL.HFDup.H3K27ac_broad_CDHFD_DB_xxx_H3K4me3_CD_HFD.bed", header=F),
enhancers_HFDdown <- read.delim("results/Epigenome_analysis/Final_enhancer_files/enh.9.HFDdown.FINAL.H3K27ac_broad_CDHFD_DB_xxx_H3K4me1_minus_me3.sort.bed", header=F),
promoters_HFDdown <- read.delim("results/Epigenome_analysis/Final_promoter_files/prom.24.FINAL.HFDdown.H3K27ac_broad_CDHFD_DB_xxx_H3K4me3_CD_HFD.bed", header=F))

names(import_list) <- c("promoters_genomewide", "enhancers_genomewide", "enhancers_allDB", "promoters_allDB", "enhancers_HFDup", "promoters_HFDup", "enhancers_HFDdown", "promoters_HFDdown")

##Filter out all locations which are NOT on a chromosome. 
import_list2 <- list()
for (i in names(import_list)) {
object <- import_list[[i]] # Save the respective dfs in a non-list object, and put that one into a list at the end.
object <- dplyr::filter(object, grepl("chr", V1))
import_list2[[i]] <- object
} 


#listEnsemblArchives()
mart <- useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl", host = "https://sep2019.archive.ensembl.org")
# use the getAnnotation function to obtain the TSS
annoData <- getAnnotation(mart, featureType = "TSS")

# Annotate the peak files.
peaks_annotated <- list()
for (i in names(import_list2)) {
object <- import_list2[[i]]
colnames(object) <- c("chrom", "start", "end")
nrow(object)
object2 <- makeGRangesFromDataFrame(object,  start.field = "start", end.field = "end", ignore.strand = T) 


#Give ranges numeric names in order
names(object2) <- c(1:length(object2))

#Annotate granges with the nearest TSS
object3 <- annotatePeakInBatch(object2, 
                               AnnotationData=annoData, 
                               featureType = "TSS",
                               output="nearestLocation",
                               PeakLocForDistance = "start")

object3 <- as.data.frame(object3)
peaks_annotated[[i]] <- object3
} 

# For the promoters, remove the duplicated gene names and only allow the closest peak to a given gene. The annotation does not take multiple TSS per gene into consideration.

library(data.table)

peaks_annotated[["promoters_genomewide"]] <- peaks_annotated[["promoters_genomewide"]] %>% dplyr::group_by(feature)
peaks_annotated[["promoters_genomewide"]] <- setDT(peaks_annotated[["promoters_genomewide"]])[order(shortestDistance), head(.SD, 1) , feature]

peaks_annotated[["promoters_HFDdown"]] <- peaks_annotated[["promoters_HFDdown"]] %>% dplyr::group_by(feature)
peaks_annotated[["promoters_HFDdown"]] <- setDT(peaks_annotated[["promoters_HFDdown"]])[order(shortestDistance), head(.SD, 1) , feature]

peaks_annotated[["promoters_HFDup"]] <- peaks_annotated[["promoters_HFDup"]] %>% dplyr::group_by(feature)
peaks_annotated[["promoters_HFDup"]] <- setDT(peaks_annotated[["promoters_HFDup"]])[order(shortestDistance), head(.SD, 1) , feature]

peaks_annotated[["promoters_allDB"]] <- peaks_annotated[["promoters_allDB"]] %>% dplyr::group_by(feature)
peaks_annotated[["promoters_allDB"]] <- setDT(peaks_annotated[["promoters_allDB"]])[order(shortestDistance), head(.SD, 1) , feature]

saveRDS(peaks_annotated, "results/Epigenome_analysis/annotated_diffbind_and_genomewide_promoters_enhancers.rds")
```

# Export the bedfiles, which are required for further analyses
```{r}
peaks_annotated_bed <- list()
for (i in names(peaks_annotated)) {

  peaks_annotated_bed[[i]] <- peaks_annotated[[i]] %>% dplyr::select("chrom"=seqnames, start, end) 
  
write.table(peaks_annotated_bed[[i]], paste0("results/Epigenome_analysis/",i, "_", as.character(length(unique(peaks_annotated[[i]]$peak))), "_", "anno.bed") ,quote = F, row.names = F, sep="\t", col.names = F)
  
}
```

# Summary statistics about distance to nearest TSS to evaluate how well the enhancers/promoters were determined.
```{r}

# Print the distance metrics
distance_metics <- list()
for (i in names(peaks_annotated)) {
  
  print(paste("The median / mean / min / max for", i, "are:"))
print(
summary(peaks_annotated[[i]]$shortestDistance))
  distance_metics[[i]] <- summary(peaks_annotated[[i]]$shortestDistance)
}

# Add a column to identify the regions
promoters_anno <- peaks_annotated$promoters_allDB %>% mutate(element = "promoters")
enhancers_anno <- peaks_annotated$enhancers_allDB %>% mutate(element = "enhancers")

promoters_genomewide_anno <- peaks_annotated$promoters_genomewide %>% mutate(element = "promoters")
enhancers_genomewide_anno <- peaks_annotated$enhancers_genomewide %>% mutate(element = "enhancers")

# Row-bind the dataframes together
combined_DBsites_anno <- rbind(promoters_anno, enhancers_anno)
combined_allSites_anno <- rbind(promoters_genomewide_anno, enhancers_genomewide_anno)

combined_DBsites_anno$shortestDistance <- combined_DBsites_anno$shortestDistance +1
combined_allSites_anno$shortestDistance <- combined_allSites_anno$shortestDistance +1

# Plot the histograms
ggplot(combined_DBsites_anno)+
  geom_histogram(aes(x=log10(shortestDistance), fill=factor(element)), binwidth=0.1, color="black", size = 0.2) +
  theme_classic() +
  scale_fill_manual("Diff-bound\nH3K27ac at", values = c("#235b95", "#73c6b6")) +
  theme(text=element_text(size=15)) +
  xlab("shortest distance to TSS (log10(bp))") 
  
ggplot(combined_allSites_anno)+
  geom_histogram(aes(x=log10(shortestDistance), fill=factor(element)), binwidth=0.1, color="black", size = 0.2) +
  theme_classic() +
  scale_fill_manual("All putative", values = c("#235b95", "#73c6b6")) +
  theme(text=element_text(size=15)) +
    geom_vline(xintercept = log10(11200), color="#235b55") +
  geom_vline(xintercept = log10(245), color="#235c10") +
  xlab("shortest distance to TSS (log10(bp))") 
```
# Make a stacked bar plot with the distances. 
```{r}
# set up cut-off values 
breaks <- c(0,1000,20000,100000,10000000)
# specify interval/bin labels
tags <- c("[0-1000)","[1000-20000)", "[20000-100000)", "[100000-10000000)")
# bucketing values into bins
promoters_anno_bins <- cut(peaks_annotated$promoters_allDB$shortestDistance, 
                  breaks=breaks, 
                  include.lowest=TRUE, 
                  right=FALSE, 
                  labels=tags)

enhancers_anno_bins <- cut(peaks_annotated$enhancers_allDB$shortestDistance, 
                  breaks=breaks, 
                  include.lowest=TRUE, 
                  right=FALSE, 
                  labels=tags)

promoters_genomewide_anno_bins <- cut(peaks_annotated$promoters_genomewide$shortestDistance, 
                  breaks=breaks, 
                  include.lowest=TRUE, 
                  right=FALSE, 
                  labels=tags)

enhancers_genomewide_anno_bins <- cut(peaks_annotated$enhancers_genomewide$shortestDistance, 
                  breaks=breaks, 
                  include.lowest=TRUE, 
                  right=FALSE, 
                  labels=tags)

set.seed(500)
promoters_genomewide_anno_random_182 <- promoters_genomewide_anno[sample(nrow(promoters_genomewide_anno), size=182), ]
promoters_genomewide_anno_random_182_bins <- cut(promoters_genomewide_anno_random_182$shortestDistance, 
                  breaks=breaks, 
                  include.lowest=TRUE, 
                  right=FALSE, 
                  labels=tags)

enhancers_genomewide_anno_random_1816 <- enhancers_genomewide_anno[sample(nrow(enhancers_genomewide_anno), size=1816), ]
promoters_genomewide_anno_random_1816_bins <- cut(enhancers_genomewide_anno_random_1816$shortestDistance, 
                  breaks=breaks, 
                  include.lowest=TRUE, 
                  right=FALSE, 
                  labels=tags)

# inspect bins. The occurence of each bin is counted.
summary_distance <- data.frame(prom_DB = summary(promoters_anno_bins),
           enha_DB = summary(enhancers_anno_bins),
           prom_all = summary(promoters_genomewide_anno_bins),
           prom_all_sub = summary(promoters_genomewide_anno_random_182_bins),
           enha_all = summary(enhancers_genomewide_anno_bins),
           enha_all_sub = summary(promoters_genomewide_anno_random_1816_bins))

# Getting the colsums for number of elements
summary_distance.mat <- as.matrix(summary_distance)
colsums <- colSums(summary_distance[1:6])

# .. and normalize to the colSums to get percentages
summary_distance.mat <- as.data.frame(round((
  sweep(summary_distance.mat,2,colsums, "/")*100)
  ,2))

summary_distance <- tibble::rownames_to_column(summary_distance.mat)

# long format needed for plotting
summary_distance_long <- pivot_longer(summary_distance, cols = 2:7) %>%
  group_by(name) 

# Replacing the brackets
summary_distance_long$rowname <- gsub(pattern="\\[", "", summary_distance_long$rowname)
summary_distance_long$rowname <- gsub(pattern="\\)", "", summary_distance_long$rowname)

# Setting the order for plotting
order_dist <- c("0-1000", "1000-20000", "20000-100000", "100000-10000000")
order_element <- c("prom_DB", "prom_all", "prom_all_sub", "enha_DB", "enha_all", "enha_all_sub")

# Plotting the stacked bar plots
ggplot(summary_distance_long) +
  geom_col(aes(y=value, x=factor(name, levels=order_element), fill=factor(rowname, levels=order_dist))) +
  scale_fill_manual("Distance\nfrom TSS (bp)", values=c("#D9D9D9","#B3B3B3", "#666666","#000000")) +
  ylab("Percentage of elements") +
  xlab("") + 
  theme(text=element_text(size=15)) +
  theme_classic()
```

```{r}
sessionInfo()
```

