---
title: 'Hepatoprotective effects of systemic ER activation'
subtitle: 'BulkRNAseq - Transcriptome molecular signatures'
author: 'Christian Sommerauer & Carlos Gallardo'
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    highlight: tango
---

```{r setup, include=FALSE}
# set path
path <- strsplit(dirname(rstudioapi::getActiveDocumentContext()$path), '/')[[1]]
path <- path[-c((length(path)-0):length(path))]
path <- paste(path, collapse = '/')
knitr::opts_knit$set(root.dir = path)
knitr::opts_chunk$set(echo = T, message = F, warning = F, dev = 'png')

```

```{r}
# source and library import
source('code/00_helper_functions.R')
library(tidyverse)
library(DESeq2)

```

```{r}
# color palettes
colPals <- list()
colPals$conditions <- setNames(c('#E98BB6', '#B02262', '#7F9AD7', '#2A2F72', '#7DC7D1', '#339ACD', '#356FB5', '#2B40AB'),
                               c('CDf', 'HFDf', 'CDm', 'HFDm', 'DPN', 'DIP', 'E2', 'PPT'))
colPals$RdBu <- rev(RColorBrewer::brewer.pal(n=11, name = 'RdBu'))
colPals$UpDown <- setNames(colPals$RdBu[c(10,2)],
                           c('up', 'down'))

```


# Load data

```{r}
# consensus differentially expressed genes
DEGs <- readRDS('results/bulkRNAseq_mmus_DEGs.rds')

# raw counts RNAseq
raw_counts <- read.table(
  file = 'data/bulkRNAseq_mmus_rawcounts.tsv',
  stringsAsFactors = FALSE,
  sep = '\t',
  header = TRUE) %>%
  dplyr::filter(geneID %in% DEGs$unfilt$CDfVsCDm$ensembl_gene_id) %>% 
  tibble::column_to_rownames('geneID') %>% 
  as.matrix() 

# gene lengths
gene_len <- read.table(
  file = 'data/bulkRNAseq_mmus_gene_lengths.tsv',
  stringsAsFactors = FALSE,
  sep = '\t',
  header = TRUE) %>% 
  dplyr::filter(geneID %in% DEGs$unfilt$CDfVsCDm$ensembl_gene_id)

# design RNAseq
design_meta <- read.table(
  file = 'data/bulkRNAseq_mmus_design.tsv',
  stringsAsFactors = FALSE,
  sep = '\t',
  header = TRUE)

# ensembl gene annotation (Mus musculus)
gene_ann <- read.table(
  file = 'data/ensembl_mmus_sep2019_annotation.tsv',
  stringsAsFactors = FALSE,
  sep = '\t',
  header = TRUE,
  fill = FALSE,
  quote = '') %>% 
  dplyr::filter(ensembl_gene_id %in% DEGs$unfilt$CDfVsCDm$ensembl_gene_id) %>% 
  dplyr::arrange(factor(ensembl_gene_id, levels = rownames(raw_counts)))

```


# Principal component analysis (PCA)

```{r}
pca_res <- DESeq2::DESeqDataSetFromMatrix(countData = raw_counts, 
                                  colData = design_meta, 
                                  design = ~0 + condition) %>% 
  DESeq2::estimateSizeFactors() %>% 
  DESeq2::DESeq() %>% 
  DESeq2::vst(blind = FALSE) %>% 
  assay() %>% 
  doPCA()

df <- data.frame(PC1 = pca_res$pcs$PC1,
                 PC2 = pca_res$pcs$PC2,
                 condition = design_meta$condition)

ggplot(df, aes(x=PC1, y=PC2, fill=condition),) +
  geom_point(shape=21, size=5, stroke=0.5, color='black') +
  scale_fill_manual(values = alpha(colPals$conditions, 0.9)) +
  scale_x_continuous(expand = expansion(mult = c(.1, .1))) +
  scale_y_continuous(
    expand =expansion(mult = c(.1, .1))) +
  xlab(paste0('PC1 (', round(pca_res$perc_var[1]), '%)')) +
  ylab(paste0('PC2 (', round(pca_res$perc_var[2]), '%)')) +
  theme_bw()

```


# Differentially expressed genes (DEGs)

```{r}
up <- lapply(DEGs$filt, function(x) sum(x$log2FoldChange>0)) %>% unlist()
down <- lapply(DEGs$filt, function(x) sum(x$log2FoldChange<0)) %>% unlist()

df <- data.frame(comparison=factor(rep(names(DEGs$filt),2), levels=names(DEGs$filt)),
                 state=c(rep('up', length(up)), rep('down', length(down))),
                 value=c(up, down*-1))

ggplot(df, aes(x=comparison, y=value, fill=state, label=value)) +
  geom_hline(yintercept = 0, linetype='solid', size=0.5) +
  geom_bar(color='black', size=0.5, width=0.8, position='stack', stat='identity') +
  geom_text(size=6) +
  scale_fill_manual(values = colPals$UpDown) +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(limits = c(-500,500), labels = c(500,250,0,250,500)) +
  coord_flip() +
  theme_bw()

```


# Filter and normalize RNAseq data

```{r}
RNAseq <- list()

# remove outlier sample PPT_HFD_male_4
RNAseq$counts <- raw_counts %>% 
  as.data.frame() %>% 
  dplyr::select(-PPT_HFD_male_4)

RNAseq$design_meta <- design_meta %>% 
  dplyr::filter(sample != 'PPT_HFD_male_4')

RNAseq$annotation <- gene_ann %>% 
  dplyr::rename(geneID = ensembl_gene_id) %>% 
  dplyr::left_join(gene_len, by = 'geneID')

RNAseq$cpm <- RNAseq$counts %>% 
  normalizeData(method = 'CPM')

RNAseq$tpm <- RNAseq$counts %>% 
  normalizeData(len = RNAseq$annotation$length, method = 'TPM')

```


# Transcriptome-wide signal-to-noise ratios (tSNR)

```{r, fig.width=4, fig.height=6}
df <- RNAseq$tpm %>% 
  scaleData(method = 'zscore') %>% 
  tSNR(group.lbls = RNAseq$design_meta$condition) %>%
  tibble::rownames_to_column(var = 'X') %>% 
  tidyr::pivot_longer(cols = dplyr::everything()[-1], names_to = 'Y') %>% 
  tidyr::unite(col = 'comparison', X, Y, sep = 'Vs') %>% 
  dplyr::filter(comparison %in% names(DEGs$filt)) %>% 
  dplyr::mutate(comparison=factor(comparison, levels = names(DEGs$filt)))

ggplot(df, aes(x=comparison, y=value)) +
  geom_line(group=1, size=1.2) +
  geom_point(shape=21, size=3, stroke=1.5, color='black', fill='white') +
  geom_hline(yintercept = 1, linetype='dashed', size=1.2, color='#EF2126') +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(limits = c(1,3), breaks = c(1,2,3)) +
  coord_flip() +
  theme_bw()

```

# Plot Venn Diagram (Fig S1E)
```{r}
library(tidyverse)
library(VennDiagram)

CDf_HFDf <- DEGs$filt$CDfVsHFDf
CDm_HFDm <- DEGs$filt$CDmVsHFDm

grid.newpage()
myCol <- c("#B02262", "#2A2E72")
venn.plot <- venn.diagram(
  x = list(CDf_HFDf$ensembl_gene_id, CDm_HFDm$ensembl_gene_id),
  category.names = c("CDf_HFDf\n(n=327)", "CDm_HFDm \n(n=714)"),
  filename = NULL,
  output=T,
  lwd = 2,
  lty = 'blank',
  fill = myCol,
    cex = 2.5,
        fontface = "bold",
        fontfamily = "sans",
  cat.cex = 1.2,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.fontfamily = "sans",
  cat.pos = c(-10, 10),
        cat.dist = c(0.08, 0.08))

grid.draw(venn.plot)

# Make intersection to extract the genes names from Venn.
intersections <- list("", "", "")
names(intersections) <- c("shared_genes", "CDf_HFDf_unique", "CDm_HFDm_unique")

intersections[[1]] <- intersect(CDf_HFDf$ensembl_gene_id,CDm_HFDm$ensembl_gene_id)
intersections[[2]] <- setdiff(CDf_HFDf$ensembl_gene_id, intersections$shared_genes)
intersections[[3]] <- setdiff(CDm_HFDm$ensembl_gene_id, intersections$shared_genes)
```

# Gene Ontologies (Fig S1F,G,H)
```{r}
library(clusterProfiler)
background <- DEGs$unfilt$CDfVsCDm$ensembl_gene_id 
options(connectionObserver = NULL) # workaround due to a bug
library(org.Mm.eg.db)

GO_list <- list(GO.results = list(),
                GO.top8 = list(),
                term.order.plotting = list())

for (i in 1:3) { 
GO_list$GO.results[[i]] <- enrichGO(gene = intersections[[i]],
                keyType       = 'ENSEMBL',
                OrgDb         = org.Mm.eg.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05,
                minGSSize     = 3,
                readable      = TRUE,
                universe = background)
head(GO_list$GO.results[[i]])

name.me <- c("shared_genes", "CDf_HFDf_unique", "CDm_HFDm_unique")

GO_list$GO.top8[[i]] <- GO_list$GO.results[[i]]@result %>% filter(p.adjust<0.05) %>% mutate(GeneSet = name.me[i]) %>% dplyr::arrange(p.adjust) %>% dplyr::slice(1:8)

GO_list$term.order.plotting[[i]] <- GO_list$GO.top8[[i]] %>% dplyr::pull("Description")

print(ggplot(GO_list$GO.top8[[i]], aes(x=Count, y=factor(Description, levels=rev(GO_list$term.order.plotting[[i]])))) +
  geom_col(aes(fill=-log10(p.adjust))) +
  theme_classic() +
  xlab("") +
  ylab("") +
  ggtitle(paste(unique(GO_list$GO.top8[[i]]$GeneSet))) +
  scale_fill_gradient(low="grey", high= "black", limits=c(1,13)))

} 
```

# Plot volcanos (Fig S1I,J)
```{r}
library(tidyverse)
library(ggrepel)

# Load dataframes from rds object.
## DEG in CD comparison
CDf_CDm <- DEGs$filt$CDfVsCDm

## All genes in given comparisons with fold-changes
CDf_CDm_unfilt <- DEGs$unfilt$CDfVsCDm
HFDf_HFDm_unfilt <- DEGs$unfilt$HFDfVsHFDm

# CD comparison
## Maximum significance and fold-change are capped here.
CDf_CDm$padj[CDf_CDm$padj < 10e-30] <- 10e-30
CDf_CDm$log2FoldChange[CDf_CDm$log2FoldChange < (-8)] <- (-8)
CDf_CDm$log2FoldChange[CDf_CDm$log2FoldChange > (8)] <- (8)

CDf_CDm_unfilt$padj[CDf_CDm_unfilt$padj < 10e-30] <- 10e-30 
CDf_CDm_unfilt$log2FoldChange[CDf_CDm_unfilt$log2FoldChange < (-8)] <- (-8)
CDf_CDm_unfilt$log2FoldChange[CDf_CDm_unfilt$log2FoldChange  > (8)] <- (8)

# HFD comparison
## We do not need the HFD DEG, because we will plot the CDf vs CDm DEG with the HFD fold-changes

HFDf_HFDm_unfilt$padj[HFDf_HFDm_unfilt$padj < 10e-30] <- 10e-30
HFDf_HFDm_unfilt$log2FoldChange[HFDf_HFDm_unfilt$log2FoldChange < (-8)] <- (-8)
HFDf_HFDm_unfilt$log2FoldChange[HFDf_HFDm_unfilt$log2FoldChange  > (8)] <- (8)

# Create up and downregulated genes in respective comparisons (larger/smaller than 0 is sufficient because the tables are already filtered for log2(1.75) and FDR < 0.05)
male_biased_CD <- filter(CDf_CDm, log2FoldChange < 0)
nrow(male_biased_CD)
female_biased_CD <- filter(CDf_CDm, log2FoldChange > 0)
nrow(female_biased_CD)
```

```{r}
# These genes will be plotted in both, HFD and CD volcano plots, as sex-biased gene set.
male_biased_CD_vector <- filter(CDf_CDm, log2FoldChange < 0) %>% dplyr::pull("external_gene_name")
female_biased_CD_vector <- filter(CDf_CDm, log2FoldChange > 0) %>% dplyr::pull("external_gene_name")
# Filter the HFD table for genes that are sex-biased in CD.
male_biased_CD_in_HFDbckgrd <- HFDf_HFDm_unfilt %>% filter(external_gene_name%in%male_biased_CD_vector)
female_biased_CD_in_HFDbckgrd <- HFDf_HFDm_unfilt %>% filter(external_gene_name%in%female_biased_CD_vector)
```

```{r}
ggplot(CDf_CDm_unfilt) +
  geom_point(data = CDf_CDm_unfilt,
    aes(x = log2FoldChange, y = -log10(padj)),
    color = "grey",
    alpha = 0.3,
    cex = 1.5) +
  geom_point(data = male_biased_CD,
    aes(x = log2FoldChange, y = -log10(padj)),
    color = "#3498db", # #3498db is blue
    alpha = 0.8,
    cex = 2) +
  geom_point(data = female_biased_CD,
    aes(x = log2FoldChange, y = -log10(padj)),
    color =  "#e74c3c" ,
    alpha = 0.8,
    cex = 2) +
  theme_classic() +
  theme(axis.text = element_text(size=20),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size=15)) +
  scale_x_continuous(limits = c(-8.1, 8.1), breaks =c(-8, -4, 0, 4, 8)) +
  xlab("fold change (log2)") +
  ylab("padj (-log10)") +
  geom_vline(xintercept = 0.807,
    col = "#2e4053",
    linetype = "dotted",
    size = 1.5) +
  geom_vline(xintercept = -0.807,
    col = "#2e4053",
    linetype = "dotted",
    size = 1.5) +
  geom_hline(yintercept = -log10(0.05),
    col = "#2e4053",
    linetype = "dotted",
    size = 1.5) +
  ggtitle("Diff. expressed sex-biased genes in control diet\n")
```
```{r}
ggplot(HFDf_HFDm_unfilt) +
  geom_point(data = HFDf_HFDm_unfilt,
    aes(x = log2FoldChange, y = -log10(padj)),
    color = "grey",
    alpha = 0.3,
    cex = 1.5) +
  geom_point(data = male_biased_CD_in_HFDbckgrd,
    aes(x = log2FoldChange, y = -log10(padj)),
    color = "#3498db",
    alpha = 0.8,
    cex = 2) +
  geom_point(data = female_biased_CD_in_HFDbckgrd,
    aes(x = log2FoldChange, y = -log10(padj)),
    color =  "#e74c3c" ,
    alpha = 0.8,
    cex = 2) +
  theme_classic() +
  theme(axis.text = element_text(size=20),
        axis.title.x = element_text(size=15),
        axis.title.y = element_text(size=15)) +
  xlab("fold change (log2)") +
  ylab("padj (-log10)") +
  scale_x_continuous(limits = c(-8.1, 8.1), breaks =c(-8, -4, 0, 4, 8)) +
  geom_vline(xintercept = 0.807,
    col = "#2e4053",
    linetype = "dotted",
    size = 1.5) +
  geom_vline(xintercept = -0.807,
    col = "#2e4053",
    linetype = "dotted",
    size = 1.5) +
  geom_hline(yintercept = -log10(0.05),
    col = "#2e4053",
    linetype = "dotted",
    size = 1.5) +
  ggtitle("Fold-changes of Sex-biased (in CD) genes in \n HFD mice")
```

# Export filtered and normalized RNAseq data

```{r}
saveRDS(RNAseq, file = 'results/bulkRNAseq_mmus_data_filt_norm.rds')
```


```{r}
sessionInfo()

```
