---
title: 'Hepatoprotective effects of systemic ER activation'
subtitle: 'BulkRNAseq - Transcriptome molecular signatures'
author: 'Christian Sommerauer & Carlos Gallardo'
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: flatly
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: false
    highlight: tango
---

```{r setup, include=FALSE}
# set path
path <- strsplit(dirname(rstudioapi::getActiveDocumentContext()$path), '/')[[1]]
path <- path[-c((length(path)-0):length(path))]
path <- paste(path, collapse = '/')
knitr::opts_knit$set(root.dir = path)
knitr::opts_chunk$set(echo = T, message = F, warning = F)

```

```{r}
# source and library import
source("code/00_helper_functions.R")
library(tidyverse)
library(DESeq2)

```

```{r}
# color palettes
colPals <- list()
colPals$conditions <- setNames(c('#E98BB6', '#B02262', '#7F9AD7', '#2A2F72', '#7DC7D1', '#339ACD', '#356FB5', '#2B40AB'),
                               c('CDf', 'HFDf', 'CDm', 'HFDm', 'DPN', 'DIP', 'E2', 'PPT'))
colPals$RdBu <- rev(RColorBrewer::brewer.pal(n=11, name = 'RdBu'))
colPals$UpDown <- setNames(colPals$RdBu[c(10,2)],
                           c('up', 'down'))

```


# Load data

```{r}
# consensus differentially expressed genes
DEGs <- readRDS('data/bulkRNAseq_mmus_DEGs.rds')

# raw counts RNAseq
raw_counts <- read.table(
  file = 'data/bulkRNAseq_mmus_rawcounts.tsv',
  stringsAsFactors = FALSE,
  sep = "\t",
  header = TRUE) %>%
  dplyr::filter(geneID %in% DEGs$unfilt$CDfVsCDm$ensembl_gene_id) %>% 
  tibble::column_to_rownames("geneID") %>% 
  as.matrix() 

# gene lengths
gene_len <- read.table(
  file = 'data/bulkRNAseq_mmus_gene_lengths.tsv',
  stringsAsFactors = FALSE,
  sep = "\t",
  header = TRUE) %>% 
  dplyr::filter(geneID %in% DEGs$unfilt$CDfVsCDm$ensembl_gene_id)

# design RNAseq
design_meta <- read.table(
  file = 'data/bulkRNAseq_mmus_design.tsv',
  stringsAsFactors = FALSE,
  sep = "\t",
  header = TRUE)

# ensembl gene annotation (Mus musculus)
gene_ann <- read.table(
  file = 'data/ensembl_mmus_sep2019_annotation.tsv',
  stringsAsFactors = FALSE,
  sep = "\t",
  header = TRUE,
  fill = FALSE,
  quote = "") %>% 
  dplyr::filter(ensembl_gene_id %in% DEGs$unfilt$CDfVsCDm$ensembl_gene_id) %>% 
  dplyr::arrange(factor(ensembl_gene_id, levels = rownames(raw_counts)))

```


# Principal component analysis (PCA)

```{r}
pca_res <- DESeq2::DESeqDataSetFromMatrix(countData = raw_counts, 
                                  colData = design_meta, 
                                  design = ~0 + condition) %>% 
  DESeq2::estimateSizeFactors() %>% 
  DESeq2::DESeq() %>% 
  DESeq2::vst(blind = FALSE) %>% 
  assay() %>% 
  doPCA()

df <- data.frame(PC1 = pca_res$pcs$PC1,
                 PC2 = pca_res$pcs$PC2,
                 condition = design_meta$condition)

ggplot(df, aes(x=PC1, y=PC2, fill=condition),) +
  geom_point(shape=21, size=5, stroke=0.5, color="black") +
  scale_fill_manual(values = alpha(colPals$conditions, 0.9)) +
  scale_x_continuous(expand = expansion(mult = c(.1, .1))) +
  scale_y_continuous(
    expand =expansion(mult = c(.1, .1))) +
  xlab(paste0('PC1 (', round(pca_res$perc_var[1]), '%)')) +
  ylab(paste0('PC2 (', round(pca_res$perc_var[2]), '%)')) +
  theme_bw()

```


# Differentially expressed genes (DEGs)

```{r}
up <- lapply(DEGs$filt, function(x) sum(x$log2FoldChange>0)) %>% unlist()
down <- lapply(DEGs$filt, function(x) sum(x$log2FoldChange<0)) %>% unlist()

df <- data.frame(comparison=factor(rep(names(DEGs$filt),2), levels=names(DEGs$filt)),
                 state=c(rep('up', length(up)), rep('down', length(down))),
                 value=c(up, down*-1))

ggplot(df, aes(x=comparison, y=value, fill=state, label=value)) +
  geom_hline(yintercept = 0, linetype="solid", size=0.5) +
  geom_bar(color="black", size=0.5, width=0.8, position="stack", stat="identity") +
  geom_text(size=6) +
  scale_fill_manual(values = colPals$UpDown) +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(limits = c(-500,500), labels = c(500,250,0,250,500)) +
  coord_flip() +
  theme_bw()

```


# Filter and normalize RNAseq data

```{r}
RNAseq <- list()

# remove outlier sample PPT_HFD_male_4
RNAseq$counts <- raw_counts %>% 
  as.data.frame() %>% 
  dplyr::select(-PPT_HFD_male_4)

RNAseq$design_meta <- design_meta %>% 
  dplyr::filter(sample != 'PPT_HFD_male_4')

RNAseq$annotation <- gene_ann %>% 
  dplyr::rename(geneID = ensembl_gene_id) %>% 
  dplyr::left_join(gene_len, by = 'geneID')

RNAseq$cpm <- RNAseq$counts %>% 
  normalizeData(method = 'CPM')

RNAseq$tpm <- RNAseq$counts %>% 
  normalizeData(len = RNAseq$annotation$length, method = 'TPM')

```


# Transcriptome-wide signal-to-noise ratios (tSNR)

```{r, fig.width=4, fig.height=6}
df <- RNAseq$tpm %>% 
  scaleData(method = 'zscore') %>% 
  tSNR(group.lbls = RNAseq$design_meta$condition) %>%
  tibble::rownames_to_column(var = 'X') %>% 
  tidyr::pivot_longer(cols = dplyr::everything()[-1], names_to = 'Y') %>% 
  tidyr::unite(col = 'comparison', X, Y, sep = 'Vs') %>% 
  dplyr::filter(comparison %in% names(DEGs$filt)) %>% 
  dplyr::mutate(comparison=factor(comparison, levels = names(DEGs$filt)))

ggplot(df, aes(x=comparison, y=value)) +
  geom_line(group=1, size=1.2) +
  geom_point(shape=21, size=3, stroke=1.5, color='black', fill='white') +
  geom_hline(yintercept = 1, linetype="dashed", size=1.2, color="#EF2126") +
  scale_x_discrete(limits = rev) +
  scale_y_continuous(limits = c(1,3), breaks = c(1,2,3)) +
  coord_flip() +
  theme_bw()

```


# TBA Fig. S1E-J

```{r}
# @christian please add code for figures S1E-J

```


# Export filtered and normalized RNAseq data

```{r}
saveRDS(RNAseq, file = 'data/bulkRNAseq_mmus_data_filt_norm.rds')

```


```{r}
sessionInfo()

```
