
---
title: 'Hepatoprotective effects of systemic ER activation'
subtitle: 'Spheroid RNA-seq - Differential expression analysis'
author: 'Christian Sommerauer & Carlos Gallardo'
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    highlight: tango
---

```{r setup, include=FALSE}
# set path
path <- strsplit(dirname(rstudioapi::getActiveDocumentContext()$path), '/')[[1]]
path <- path[-c((length(path)-0):length(path))]
path <- paste(path, collapse = '/')
knitr::opts_knit$set(root.dir = path)
knitr::opts_chunk$set(echo = T, message = F, warning = F, dev = 'png')

```

# Load packages
```{r}
library(tidyverse)
library(DESeq2)
library(biomaRt)
```

# Load featureCount table
```{r}
count_table <- read.table("data/raw_fq_featurecounts_inhibitors_spheroids.txt", header=T) %>% drop_na()

names(count_table) <- gsub("hisat2_aligned\\.|\\.bam", "", names(count_table)) 

count_table.2 <- count_table  %>% tibble::column_to_rownames("Geneid") %>% dplyr::select(2:9) 
head(count_table.2)
```
# Design metatables
```{r}
sample_name_inhibitors <- names(count_table.2)
replicate_inhibitors <- c(1,2,3,1,2,1,2,3)
Condition_inhibitors <- c(rep("control",3), rep("TEADap",2), rep("TEADsf",3))
meta_table_inhibitors <- data.frame(Sample=sample_name_inhibitors, Condition=Condition_inhibitors, Replicate=replicate_inhibitors)
write.table(meta_table_inhibitors, "results/spheroid_meta_table.txt", sep="\t", quote=F)
```

# Export TPM-normalized count table
```{r}
# normalize to TPM
source("code/00_helper_functions.R")
count_table_TPM <- normalizeData(x=count_table.2, method = "TPM", len = count_table$Length) 

count_table_TPM_mean <- groupTransform_nontibble(x=count_table_TPM, 
               group.lbls = meta_table_inhibitors$Condition,
               FUN=function(x) apply(x, 1, mean)) %>% 
               tibble::rownames_to_column("ensembl_gene_id")

# move the rownames to column
count_table_TPM <- count_table_TPM %>% 
              tibble::rownames_to_column("ensembl_gene_id")

# Add the external_gene_names 
#listEnsemblArchives()
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", host="https://oct2022.archive.ensembl.org") #, host = "https://oct2022.archive.ensembl.org"
annoData <- getBM(attributes=c("ensembl_gene_id","external_gene_name", "description"),mart=mart)

export_TPM_normTables <- list(count_table_TPM=count_table_TPM, 
                              count_table_TPM_mean=count_table_TPM_mean)

export_TPM_normTables_single <- inner_join(export_TPM_normTables$count_table_TPM, y = annoData, by = 'ensembl_gene_id')
export_TPM_normTables_mean <- inner_join(export_TPM_normTables$count_table_TPM_mean, y = annoData, by = 'ensembl_gene_id')

write.table(export_TPM_normTables_single, "results/spheroid_TPM_norm_counts.txt", sep="\t", quote=F, row.names = F)
write.table(export_TPM_normTables_mean, "results/spheroid_TPM_norm_counts_mean.txt", sep="\t", quote=F, row.names = F)
```

# Run DESeq2 function and plot a PCA
```{r}
ds_data_inhibitors <- DESeqDataSetFromMatrix(countData = count_table.2,
                                 colData = meta_table_inhibitors,
                                 design = ~ 0 + Condition)
ds_data_inhibitors <- estimateSizeFactors(ds_data_inhibitors)
ds_data_inhibitors <- DESeq(ds_data_inhibitors)

# filtering to minimum 15 counts in at least 8 samples
ds_data_inhibitors.filt <- ds_data_inhibitors[rowSums(counts(ds_data_inhibitors, normalized=TRUE) >= 10 ) >= 2, ]

vsd_inhibitors <- vst(ds_data_inhibitors.filt)

PCA.to.plot_inhibitors <- plotPCA(vsd_inhibitors, intgroup=c("Condition")) + theme_bw() +  ggtitle("PCA of RNA-Seq Samples (DESeq2, vst)") 
PCA.to.plot_inhibitors
ggplot(PCA.to.plot_inhibitors$data, aes(x=PC1, y=PC2, fill=group), color="black") + 
  geom_point(shape=21, size=7) +
  theme_light() +
  ylab("PC2 - 14% variance explained") + # took the number from the original plotPCA function
  xlab("PC1 - 72% variance explained")
```

```{r}
DESeq2_DEGs_inhibitors <- list()
# comparisons
DESeq2_DEGs_inhibitors$unfilt <- list(
  TEADap_vs_control = results(ds_data_inhibitors.filt, contrast = c('Condition', 'TEADap', 'control')),
  TEADsf_vs_control = results(ds_data_inhibitors.filt, contrast = c('Condition', 'TEADsf', 'control')))

names(DESeq2_DEGs_inhibitors$unfilt)

# add annotation to DEG lists
DESeq2_DEGs_inhibitors$unfilt <- lapply(DESeq2_DEGs_inhibitors$unfilt, as.data.frame)
DESeq2_DEGs_inhibitors$unfilt <- lapply(DESeq2_DEGs_inhibitors$unfilt, rownames_to_column, var = 'ensembl_gene_id')

listEnsemblArchives()
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", host = "https://oct2022.archive.ensembl.org")
# use the getAnnotation function to obtain relevant features for ensembl GRCh38.p13.
annoData <- getBM(attributes=c("ensembl_gene_id","external_gene_name", "chromosome_name", "gene_biotype", "description"),mart=mart)

DESeq2_DEGs_inhibitors$unfilt <- lapply(DESeq2_DEGs_inhibitors$unfilt, inner_join, y = annoData, by = 'ensembl_gene_id')

DEG_counts_inhibitors <- vector()
for (i in 1:2) { 
DESeq2_DEGs_inhibitors$filt[[i]] <- DESeq2_DEGs_inhibitors$unfilt[[i]] %>% filter(abs(log2FoldChange) > 0.585 & padj<0.05)
DEG_counts_inhibitors[i] <- nrow(DESeq2_DEGs_inhibitors$filt[[i]])
} 

names(DESeq2_DEGs_inhibitors$filt) = names(DESeq2_DEGs_inhibitors$unfilt)

DEG_counts_inhibitors <- data.frame(Condition = names(DESeq2_DEGs_inhibitors$filt), DEG_counts = DEG_counts_inhibitors)

ggplot(DEG_counts_inhibitors, aes(x=DEG_counts, y=factor(Condition, levels=rev(names(DESeq2_DEGs_inhibitors$unfilt))))) +
  geom_col() +
  theme_classic() +
  ylab("")

saveRDS(DESeq2_DEGs_inhibitors, "results/Spheroid_inhibitors_DEGlists.rds")
```

```{r}
sessionInfo()
```

