
---
title: 'Hepatoprotective effects of systemic ER activation'
subtitle: 'Spheroid RNA-seq - Differential expression analysis'
author: 'Christian Sommerauer & Carlos Gallardo'
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    highlight: tango
---

```{r setup, include=FALSE}
# set path
path <- strsplit(dirname(rstudioapi::getActiveDocumentContext()$path), '/')[[1]]
path <- path[-c((length(path)-0):length(path))]
path <- paste(path, collapse = '/')
knitr::opts_knit$set(root.dir = path)
knitr::opts_chunk$set(echo = T, message = F, warning = F, dev = 'png')

```

# Load packages
```{r}
library(clusterProfiler)
library(tidyverse)
library(org.Hs.eg.db)
library(fgsea)
sessionInfo()
```

# Load RDS files
```{r}
inhibitor_RDS <- readRDS("results/Spheroid_inhibitors_DEGlists.rds")

universe <- inhibitor_RDS$unfilt$TEADap_vs_control %>% pull("external_gene_name")

TEADap_inh_vs_control <- inhibitor_RDS$filt$TEADap_vs_control
TEADsf_inh_vs_control <- inhibitor_RDS$filt$TEADsf_vs_control
```

```{r}
TEADap_inh_vs_control_entrez <- bitr(TEADap_inh_vs_control$ensembl_gene_id, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = "org.Hs.eg.db")
TEADsf_inh_vs_control_entrez <- bitr(TEADsf_inh_vs_control$ensembl_gene_id, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = "org.Hs.eg.db")


KEGG_TEADap_inh <- enrichKEGG(gene = TEADap_inh_vs_control_entrez$ENTREZID,
          organism = "hsa",
          keyType = "kegg",
          pAdjustMethod = "BH",
          qvalueCutoff = 0.05)
KEGG_TEADap_inh <- as.data.frame(KEGG_TEADap_inh@result)
#write.table(KEGG_TEADap_inh, "data/KEGG_TEADap_vs_control_FigS8A.txt", sep="\t", quote=F, row.names = F)

KEGG_TEADsf_inh <- enrichKEGG(gene = TEADsf_inh_vs_control_entrez$ENTREZID,
          organism = "hsa",
          keyType = "kegg",
          pAdjustMethod = "BH",
          qvalueCutoff = 0.05)
KEGG_TEADsf_inh <- as.data.frame(KEGG_TEADsf_inh@result)


#This function tells you the gene symbols in each KEGG pathway (changed by TEADap)
gimme_SYMBOL_from_KEGG <- function(table, pathway) {
 table_subset <- table %>%  filter(Description==pathway) %>% pull("geneID")
 split_entrez <- unlist(strsplit(as.character(table_subset), "/"))
 symbols <- bitr(split_entrez, fromType = "ENTREZID", toType = "SYMBOL", OrgDb = "org.Hs.eg.db")
 symbols <- symbols %>%  mutate(Pathway = pathway)
 symbols
 
}


#This function tells you the gene symbols in each KEGG pathway (changed by TEADap)
gimme_SYMBOL_from_KEGG(table=KEGG_TEADap_inh, pathway = "AMPK signaling pathway")
gimme_SYMBOL_from_KEGG(table=KEGG_TEADap_inh, pathway = "PI3K-Akt signaling pathway")

# This code exports the genes inside the top12 KEGG pathways of TEADap vs control with the corresponding human-readable symbols.
KEGG_top12_pathway_geneList <- list()
for (i in KEGG_TEADap_inh$Description[1:12]) {
  KEGG_top12_pathway_geneList[[i]] <- gimme_SYMBOL_from_KEGG(table=KEGG_TEADap_inh, pathway=i)
}

#saveRDS(KEGG_top12_pathway_geneList, "data/top12KEGGpathways_gene_content.rds")
```
# Plotting the KEGG pathways
```{r}
# Import the KEGG table (run in June 2023, KEGG version from April)
KEGG_TEADap_inh <- read.delim("data/KEGG_TEADap_vs_control_FigS8A.txt")

# Note: This following section is looped, so one can technically plot many results at once.
list_KEGG <- list(KEGG_TEADap_inh=KEGG_TEADap_inh,
                  KEGG_TEADsf_inh=KEGG_TEADsf_inh)

combined_genesets <- list()

for (i in names(list_KEGG)) {
  
combined_genesets[[i]] <- list_KEGG[[i]] %>%
               dplyr::arrange(p.adjust) %>%
               dplyr::slice(1:12)

names(combined_genesets[[i]])
order <- list()
order[[i]] <- combined_genesets[[i]] %>% dplyr::pull(Description)
# Plot and order by p.adjust
ggplot_bar <- ggplot(combined_genesets[[i]], aes(x=factor(Description, levels=order[[i]]), y=-log10(p.adjust), fill=Count)) +
  geom_col(color="black") +
  coord_flip() +
  scale_x_discrete(limits=rev) +
  #scale_y_continuous("-log10") +
  scale_fill_gradient(high = "black", low="grey") +
  theme_classic() +
  theme(text=element_text(size=20)) +
  geom_hline(yintercept = -log10(0.05), color="red", linewidth=2) +
  xlab("") +
  ggtitle(i)

date <- gsub("-", "", Sys.Date())

print(ggplot_bar)

#ggsave(filename = paste0("results/","SupplFig8A_KEGG_RNAseq_top12_", i, ".pdf"), width=10, height=7)
       
} 

```

# Gene expression trend analysis
```{r}
library(Mfuzz)
# CLUSTERING WITH ONLY Inhibitors DEG

# Clustering of expression profiles
DESeq2_DEGs_clust <- list(TEADap_inh_vs_control, TEADsf_inh_vs_control)

names(DESeq2_DEGs_clust) <- c("TEADap_inh_vs_control", "TEADsf_inh_vs_control")
  
DEGs_union <- lapply(DESeq2_DEGs_clust, function(x) x$ensembl_gene_id) %>% 
  unlist() %>% 
  unique()

# TPM - normalized table with the average TPM (computed in Spheroid DESeq2 analysis script)
nonNorm_Counts <- read.delim("results/spheroid_TPM_norm_counts_mean.txt")

TPM_Normalized <- nonNorm_Counts %>% tibble::column_to_rownames("ensembl_gene_id") %>% 
  dplyr::select(-description, -external_gene_name)
colSums(TPM_Normalized) 
 
source("code/00_helper_functions.R")

eset <- TPM_Normalized %>% 
  dplyr::filter(row.names(.) %in% DEGs_union)

# zscore data (mean=0, sd=1)
eset <- new('ExpressionSet', exprs = as.matrix(eset)) %>% 
  Mfuzz::standardise()

# estimate fuzzifier parameter for clustering
m_eset <- Mfuzz::mestimate(eset)

# determine cluster number with minimum centroid distance
Mfuzz::Dmin(eset, m = m_eset, crange = seq(2,20,1), repeats = 5)
```

```{r}
set.seed(3452)

# generate mfuzz clusters (n=4)
clusters <- mfuzz(eset, c = 2, m = m_eset)

# check correlation of cluster centroids
cor(t(clusters[[1]]))
```

```{r}
# get cluster membership values of genes
cluster_memberships <- acore(eset, cl = clusters, min.acore = 0.0)

# assign to cluster with top membership value
cluster_memberships <- do.call(rbind, 
                               lapply(seq_along(cluster_memberships), 
                                      function(x) {data.frame(CLUSTER=x, 
                                                              cluster_memberships[[x]])})) %>% 
  dplyr::mutate(CLUSTER=dplyr::recode(CLUSTER, !!!setNames(c(2,1), seq(1,2,1))))

# check number of genes per cluster
cluster_gene_counts <- table(cluster_memberships$CLUSTER)
cluster_gene_counts
```

```{r}
DEG_clusters <- list()
# extract gene profiles and cluster assignments
DEG_clusters$genes <- as.data.frame(exprs(eset)) %>% 
  tibble::rownames_to_column(var = 'geneID') %>% 
  tibble::add_column(GeneSymbol = .$geneID, .after = 'geneID') %>% 
  dplyr::mutate(GeneSymbol=dplyr::recode(GeneSymbol, 
                                         !!!setNames(inhibitor_RDS$unfilt$TEADap_vs_control$external_gene_name, 
                                                     inhibitor_RDS$unfilt$TEADap_vs_control$ensembl_gene_id))) %>% 
  merge(cluster_memberships, by.x = 'geneID', by.y = 'NAME', sort = F)

# extract cluster centroid profiles
DEG_clusters$centroids <- as.data.frame(clusters$centers) %>% 
  tibble::add_column(geneID = paste0('centroid_', c(2,1)), .before = 'control') %>% 
  tibble::add_column(GeneSymbol = paste0('centroid_', c(2,1)), .before = 'control') %>%
  dplyr::mutate(CLUSTER=c(2,1),
                MEM.SHIP=1) %>% 
  arrange(CLUSTER)

df <- DEG_clusters$genes %>% 
  dplyr::bind_rows(DEG_clusters$centroids) %>% 
  tidyr::pivot_longer(cols = c("control", "TEADsf", "TEADap"), 
                      names_to = 'condition', 
                      values_to = 'expression') %>% 
  dplyr::mutate(CLUSTER=factor(CLUSTER, levels = 1:2),
                condition=factor(condition, levels = c("control", "TEADsf", "TEADap")))

cluster_names <- c('1' = paste0("Cluster 1 (n=",cluster_gene_counts[1],")"),
                   '2' = paste0("Cluster 2 (n=",cluster_gene_counts[2],")"))

ggplot(df, aes(x=condition, y=expression, color=CLUSTER, group=geneID)) +
  geom_line(data = subset(df, !grepl('centroid', GeneSymbol)), size = 1,  color=alpha('#AEAEAE', 0.15)) +
  geom_line(data = subset(df, grepl('centroid', GeneSymbol)), size = 1.2, linetype="dotted", color="black") +
  geom_point(data = subset(df, grepl('centroid', GeneSymbol)), shape=21, size=3, stroke=1.5, fill='white', color="black") +
  #scale_color_manual(values = colPals$clusters) +
  facet_wrap(~CLUSTER, nrow = 1, labeller = as_labeller(cluster_names)) +
  theme_bw() +
  theme(strip.background = element_blank(),
        axis.text.x=element_text(angle=90, hjust=0.95, vjust=0.5))
  

#ggsave("cMeans_clustering_inhibitors_2clusters.pdf", width = 23, height=10, units="cm")
#write.table(DEG_clusters$genes, "cMeans_clustering_inhibitors_2clusters.txt", sep="\t", row.names = F, quote = F)
```

```{r}
sessionInfo()
```